src/server/debug/server/server.go

```diff
@@ -214,67 +148,364 @@ func (s *debugServer) appLogs(ctx context.Context, tw *tar.Writer) (retErr error
-		return s.handleWorkerRedirect(ctx, tw, pod, collectWorker, redirect, prefix)
+}
+
+func (s *debugServer) dump(c *client.APIClient, server debug.Debug_DumpV2Server, tasks []taskFunc) (retErr error) {
+	ctx := c.Ctx() // this context has authorization credentials we need
+	dumpRoot := filepath.Join(os.TempDir(), uuid.NewWithoutDashes())
+	if err := os.Mkdir(dumpRoot, os.ModeDir); err != nil {
+		return errors.Wrap(err, "create dump root directory")
+	}
+	defer func() {
+		retErr = multierr.Append(retErr, os.RemoveAll(dumpRoot))
+	}()
+	timeout := 25 * time.Minute // If no client-side deadline set, use 25 minutes.
+	if deadline, ok := server.Context().Deadline(); ok {
+		d := time.Until(deadline)
+		// Time out our own operations at ~90% of the client-set deadline. (22 minutes of
+		// server-side processing for every 25 minutes of debug dumping.)
+		timeout = time.Duration(22) * d / time.Duration(25)
+		var cf context.CancelFunc
+		ctx, cf = context.WithTimeout(ctx, timeout)
+		defer cf()
+	}
+	eg, ctx := errgroup.WithContext(ctx)
+	var mu sync.Mutex // to synchronize writes to the tar stream
+	dumpContent := &dumpContentServer{server: server}
+	return grpcutil.WithStreamingBytesWriter(dumpContent, func(w io.Writer) error {
+		return withDebugWriter(w, func(tw *tar.Writer) error {
+			for _, f := range tasks {
+				func(f taskFunc) {
+					eg.Go(func() (retErr error) {
+						taskCtx := ctx
+						defer func() {
+							if retErr != nil {
+								log.Error(taskCtx, fmt.Sprintf("failed dump task %q", "name"), zap.Error(retErr))
+								retErr = nil
+							}
+						}()
+						taskDir := filepath.Join(dumpRoot, uuid.NewWithoutDashes())
+						if err := os.Mkdir(taskDir, os.ModeDir); err != nil {
+							return errors.Wrap(err, "create dump task sub-directory")
+						}
+						defer func() {
+							if err := os.RemoveAll(taskDir); err != nil {
+								retErr = multierr.Append(retErr, err)
+							}
+						}()
+						// TODO(acohen4): how to best collect the individual errors????
+						if err := f(ctx, taskDir); err != nil {
+							if writeErr := writeErrorFile(taskDir, err); writeErr != nil {
+								log.Error(taskCtx, "write error", zap.Error(writeErr))
+							}
+						}
+						mu.Lock()
+						defer mu.Unlock()
+						return filepath.Walk(taskDir, func(path string, fi fs.FileInfo, err error) error {
+							if err != nil {
+								return errors.Wrapf(err, "walk path %q", path)
+							}
+							if fi.IsDir() {
+								return nil
+							}
+							dest := strings.TrimPrefix(path, taskDir)
+							return errors.Wrapf(writeTarFile(tw, dest, path, fi), "write tar file %q", dest)
+						})
+					})
+				}(f)
+			}
+			return errors.Wrap(eg.Wait(), "run dump tasks")
+		})
 	})
 }
 
+func (s *debugServer) lokiLogs(ctx context.Context, dir string, pod *debug.Pod) (retErr error) {
+	ctx, end := log.SpanContext(ctx, "collectLokiLogs")
+	defer end(log.Errorp(&retErr))
+	ctx, cf := context.WithTimeout(ctx, 10*time.Minute)
```

Comment:
These should be configurable.

Reply:
let's talk about this 👍

Reply:
We forgot to talk about this.  I think we can leave this as-is for now, but address this when we address pod/container differences in the spec.